From 751894cc7dc649323c829a9489b9a4f59292556e Mon Sep 17 00:00:00 2001
From: Balamurugan Ramar <quic_bramar@quicinc.com>
Date: Tue, 4 Jun 2024 14:38:59 +0530
Subject: [PATCH] mac80211: wifi: Fix for use-after-free in tx path

key memory is allocated and stored in two places,
  1. sta->ptk
  2. sta->fast_tx

After freeing the key memory, NULL pointer is assigned to sta->ptk but
not assigning to sta->fast_tx. Due to this, freed key memory accessed
in the tx path. To fix this issue, set tx key for old key after freed
the key memroy.

Signed-off-by: Balamurugan Ramar <quic_bramar@quicinc.com>
---
 net/mac80211/key.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/net/mac80211/key.c b/net/mac80211/key.c
index 50bcd6a..554bae9 100644
--- a/net/mac80211/key.c
+++ b/net/mac80211/key.c
@@ -532,6 +532,9 @@ static int ieee80211_key_replace(struct ieee80211_sub_if_data *sdata,
 			if (new &&
 			    !(new->conf.flags & IEEE80211_KEY_FLAG_NO_AUTO_TX))
 				_ieee80211_set_tx_key(new, true);
+			else if (old &&
+			         !(old->conf.flags & IEEE80211_KEY_FLAG_NO_AUTO_TX))
+				_ieee80211_set_tx_key(old, true);
 		} else {
 			rcu_assign_pointer(link_sta->gtk[idx], new);
 		}
-- 
2.34.1

